x86 : Memory domain implementation for x86

AdithyaBaglody opened the pull request:
This PR will show the implementation of memory domain for x86.

andrewboie commented on arch/x86/core/swap.S:
```diff
@@ -329,6 +329,33 @@ CROHandlingDone:
 	movl    %eax, _kernel_offset_to_current(%edi)
 
+#ifdef CONFIG_MEMORY_DOMAIN
+	/* Reset the mem_domain partitions, if configured.*/
```
there's no need to do this in assembly. put this logic in _x86_swap_update_page_tables()

AdithyaBaglody replied:
All these are with the old architecture. I will move these APIs into the correct locations.

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,155 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
+		 */
+		for(partition_index = 0;
+		    partition_index <= total_partitions;
+		    partition_index++){
```
space between ) and {

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,155 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
+		__ASSERT(total_partitions <= CONFIG_MAX_DOMAIN_PARTITIONS,
+			 "invalid partitions");
+
+		/* Interate over all the partitions for the given mem_domain
```
Iterate

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,155 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
+		}
+
+		/* Get the total number of partitions.
+		 * Block if the partitions exceed the max.*/
```
I wouldn't use the word "Block" here. This comment is redundant anyway, it's pretty obvious what the next two lines do

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,155 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
+}
+
+/* Reset/destroy one partition spcified in the argument of the API. */
+void x86_destroy_mem_partition(struct k_thread *thread, u32_t partition_index)
```
we don't need this function, the only time we need to update the page tables is on swap and when dropping to user mode, and in those cases we just update the entire domain

andrewboie commented on samples/mpu/mem_domain_apis_test/src/main.c:
```diff
@@ -26,13 +26,24 @@ struct k_thread app_thread_id[3];
 K_MEM_PARTITION_DEFINE(app0_parts0, app0_buf, sizeof(app0_buf),
 		       K_MEM_PARTITION_P_RW_U_RW);
+#ifdef CONFIG_X86
+K_MEM_PARTITION_DEFINE(app0_parts1, app1_buf, sizeof(app1_buf),
```
why do we need different flags for x86?
This should be the same on all arches, please explain your reasoning. we do not want to be going down a path where code that uses memory protection on ARM has to be ifdefed to work with memory protection on x86

AdithyaBaglody replied:
Andrew in our page tables we cant have the Privileged have write access and the user only read access.
If I set the page table 'U/S' bit,then whatever is in 'R/W' bit will be applicable for both user/supervisor.
I haven't come across a way to provide different access writes to user/supervisor in the same PTE. I think we have discussed this issue before.

andrewboie replied:
Ah, now I remember.
Let me think about this some more.

andrewboie replied:
this is OK for now

andrewboie commented on samples/mpu/mem_domain_apis_test/src/main.c:
```diff
@@ -76,7 +93,17 @@ void app_thread(void *dummy1, void *dummy2, void *dummy3)
 	while (1) {
 		printk("app thread %d (%p) yields.\n", id, &app_thread_id[id]);
-
+#ifdef CONFIG_x86		
```
get rid of this.
these tests need to be arch-agnostic and verify the public APIs.
shouldn't be making arch calls here

AdithyaBaglody replied:
This was during debugging. Will remove this.

andrewboie commented on samples/mpu/mem_domain_apis_test/src/main.c:
```diff
@@ -26,13 +26,24 @@ struct k_thread app_thread_id[3];
 struct k_mem_domain app_domain[2];
 
 /* the start address of the MPU region needs to align with its size */
+#ifdef CONFIG_ARM
```
we need to get rid of the hardcoded values here and have macros for users to do this

andrewboie replied:
still need to fix this.
what is this value supposed to represent? stack buffer guard size?

AdithyaBaglody replied:
yep it is the same. I can change it to use that stack guard macro. But just thinking here, the name of the macro might cause confusion.

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,155 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
+}
+
+/* Load only one parttion of the thread */
+void x86_load_mmu_mem_partition(struct k_thread *thread, u32_t partition_index)
```
you don't need this function. see comment below.

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,155 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
 }
+
+
+#ifdef CONFIG_MEMORY_DOMAIN
```
use CONFIG_USERSPACE. I'd rather see all this new code in arch/x86/thread.c

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,155 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
+#ifdef CONFIG_MEMORY_DOMAIN
+
+/* Load the required parttions of the new incomming thread */
+void x86_load_mmu_mem_domain(struct k_thread *thread)
```
you need to call this function from _arch_user_mode_enter

lpereira commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -69,6 +69,77 @@ static void _region_init(u32_t index, u32_t region_addr,
+
+static inline u32_t round_up_to_next_power_of_two(u32_t v)
+{
+	v--;
```
This function is equivalent to `1 << find_msb_set(v)`.

lpereira commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -69,6 +69,77 @@ static void _region_init(u32_t index, u32_t region_addr,
+ * This internal function is utilized by the MPU driver to parse the intent
+ * type (i.e. THREAD_STACK_REGION) and return the correct region index.
+ */
+static inline u32_t _get_region_index_by_type(u32_t type)
```
This is exactly the same as the k64f. Can both versions use the same code?

lpereira commented on kernel/mem_domain.c:
```diff
@@ -0,0 +1,304 @@
+
+static inline u8_t arch_get_max_partitions(void)
+{
+#if defined(CONFIG_ARM_CORE_MPU)
```
Shouldn't these `arch_*` functions be in arch-specific headers?

andrewboie replied:
yeah Chunlin is going to revise this, we discussed in PR #814

lpereira commented on kernel/mem_domain.c:
```diff
@@ -0,0 +1,304 @@
+{
+	unsigned int key;
+
+	if (!domain || !name || (num_parts && !parts)) {
```
Could these invariants be moved to assertions? If so, I'd move them to different assertions, so that, when they're logged, we know what argument was invalid.  Same for the test below and other functions.

andrewboie replied:
this is part of Chunlin's patch, I gave similar feedback in #814

dbkinder commented on doc/api/kernel_api.rst:
```diff
@@ -217,3 +217,18 @@ of variable-size data items.
 
 .. doxygengroup:: ring_buffer_apis
    :project: Zephyr
+   :content-only:
```
delete  :content-only:

dbkinder commented on doc/api/kernel_api.rst:
```diff
@@ -217,3 +217,18 @@ of variable-size data items.
+
+.. doxygengroup:: mem_domain_apis
+   :project: Zephyr
+   :content-only:
```
delete  :content-only:

dbkinder commented on doc/kernel/other/memory_domain.rst:
```diff
@@ -0,0 +1,189 @@
+    k_mem_domain_add_partition(&app0_domain, &app0_part1);
+
+.. note::
+    The maximum number of memory partitions is limited by the maixmum
```
s/maixmum/maximum/

dbkinder commented on doc/kernel/other/memory_domain.rst:
```diff
@@ -0,0 +1,189 @@
+==============================
+
+When defining a partition, we need to set access permission attributes
+to the partition. Since the access control of memory partitions relies on
```
change to, "relies on either an"

dbkinder reviewed (dismissed):
a few suggested changes

agross-linaro commented:
will review again after rebase

AdithyaBaglody committed:
kernel: arch interface for memory domain

Additional arch specific interfaces to handle memory domain
destroy and single partition removal.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
ARM: MPU: Arch specific memory domain APIs

Added architecture specific support for memory domain destroy
and remove partition for arm and nxp. An optimized version of
remove partition was also added.

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
x86: MMU: Memory domain implementation for x86

Added support for memory domain implementation.

GH-3852

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
x86: MMU: Configure page tables entries for memory domain in swap.

During swap the required page tables are configured. The outgoing
thread's memory domain pages are reset and the incoming thread's
memory domain is loaded. The pages are configured if userspace
is enabled and if memory domain has been initialized before
calling swap.

GH-3852

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
kernel: Arch specific memory domain APIs added

Added arch specific calls to handle memory domain destroy
and removal of partition.

GH-3852

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody committed:
samples: MPU: updated the memory domain test for x86.

GH-3852

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>

AdithyaBaglody commented:
Changes from the first patch:
1. removed chunlin's patchs from this PR.
2. updated the implementation acc to the new mem domain architecture.
3. added arch specific API's to handle mem domain destroy and removal of partition.
4. Fixed some CI failures

andrewboie commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -52,4 +52,37 @@ int _arch_mem_domain_max_partitions_get(void)
+{
+	ARG_UNUSED(domain);
+
+	arm_core_mpu_disable();
```
Why do we need these functions to do anything?
My understanding was that the additional _arch_mem_domain* APIs were needed for the x86 MMU, that the ARM implementation is already working fine.
So why aren't these no-ops for ARM?

AdithyaBaglody replied:
If we create an application on ARM and move it to x86 there will be inconsistencies between the working of the kernel itself. In ARM everything will be  handled in the swap and this is what the user might expect when he ports it to x86. But that will not be the case and he might see his application break at some point.

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,110 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
+
+		/* Get the total number of partitions*/
+		total_partitions = mem_domain->num_partitions;
+		__ASSERT(total_partitions <= CONFIG_MAX_DOMAIN_PARTITIONS,
```
why is this assertion placed here?
should be in the common mem_domain code, if anywhere

andrewboie commented on arch/x86/core/thread.c:
```diff
@@ -167,6 +167,15 @@ void _x86_swap_update_page_tables(struct k_thread *incoming,
 	/* TODO: if either thread defines different memory domains, efficiently
```
get rid of this TODO since we are implementing it

andrewboie commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -290,6 +290,24 @@ void arm_core_mpu_configure_mem_partition(u32_t part_index,
+void arm_core_mpu_remove_mem_partition(u32_t part_index)
+{
+	u32_t region_index =
+		_get_region_index_by_type(THREAD_DOMAIN_PARTITION_REGION);
```
same question as before...why do we need this??

andrewboie commented on arch/x86/core/thread.c:
```diff
@@ -167,6 +167,15 @@ void _x86_swap_update_page_tables(struct k_thread *incoming,
+	if (incoming->mem_domain_info.mem_domain !=
+	   outgoing->mem_domain_info.mem_domain){
+
+		/* There maybe inconsistencies in the page states.
```
What do you mean by "maybe inconsistencies"
Taking down the outgoing memory domain configuration is required on context switch.

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -136,3 +136,110 @@ void _x86_mmu_set_flags(void *ptr, size_t size, u32_t flags, u32_t mask)
+			}
+		}
+
+	} while (0);
```
what is this point of having all this in a do { } while (0) loop?
get rid of this

AdithyaBaglody replied:
This was placed for MISRA compliance for later on. which states only 1 return per function.

andrewboie replied:
don't it this way. use a goto if you want exactly one "return" statement.

AdithyaBaglody replied:
If I am not mistaken MISRA will complain for the 'goto' as well. So I implemented it this way.

lpereira replied:
We use `goto` a lot in our code; in fact, that's the preferred way of doing error recovery in Zephyr, as that's the one that generates smaller and more efficient code.

The rule in MISRA about the use of this statement (15.1) is only an advisory, and it has an exception to allow jumping only forward (15.2), which is what's used in Zephyr.

AdithyaBaglody replied:
Thanks for the clarification. Will go ahead with the goto implementation.

andrewboie commented on samples/mpu/mem_domain_apis_test/src/main.c:
```diff
@@ -110,7 +125,9 @@ void main(void)
 
 	while (1) {
 		printk("main thread (%p) yields.\n", k_current_get());
+
```
any reason for these extraneous whitespace changes?

chunlinhan commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -52,4 +52,37 @@ int _arch_mem_domain_max_partitions_get(void)
+	ARG_UNUSED(domain);
+
+	arm_core_mpu_disable();
+	arm_core_mpu_remove_mem_partition(partition_id);
```
Need to check if current thread belongs to this memory domain before we call arm_core_mpu_remove_mem_partition().

AdithyaBaglody replied:
oh ya good catch but I shall add it in the kernel code rather than in arch.

chunlinhan commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -52,4 +52,37 @@ int _arch_mem_domain_max_partitions_get(void)
+	ARG_UNUSED(domain);
+
+	arm_core_mpu_disable();
+	arm_core_mpu_configure_mem_domain(NULL);
```
Need to check if current thread belongs to this memory domain before we call arm_core_mpu_configure_mem_domain().

AdithyaBaglody commented:
Delta:
1. Check if the current thread's mem domain is getting destroyed then call the arch specific cleanup.
2. Converted the do{ }while(0); to goto statements.
3. And a few other review comments fixed.
